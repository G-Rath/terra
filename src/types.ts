export enum TFNodeType {
  Argument = 'argument',
  Function = 'function',
  Block = 'block',
  Map = 'map'
}

export type TFPrimitiveLiteral =
  | string //
  | number
  | boolean
  | null;

export interface TFMapLiteral {
  type: TFNodeType.Map;
  attributes: Array<[string, TFLiteralExpression]>;
}

export type TFLiteralExpression =
  | TFPrimitiveLiteral // todo: replace w/ Node
  | TFFunctionCall
  | TFLiteralExpression[] // todo: replace w/ Node
  | TFMapLiteral;

interface TFFunctionCall {
  type: TFNodeType.Function;
  name: string;
  args: unknown[];
}

export interface TFArgument {
  type: TFNodeType.Argument;
  identifier: string;
  expression: TFLiteralExpression;
}

export interface TFBlockLiteral {
  type: TFNodeType.Block;

  /**
   * The name of the block literal
   */
  name: string;
  /**
   * The arguments that make up the resource.
   */
  body: TFBlockBody;
}

export type TFBlockBody = Array<TFArgument | TFBlockLiteral>;
