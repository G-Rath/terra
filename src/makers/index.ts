export * from './makeTFLabel';
export * from './makeTFBlock';
export * from './makeTFAttribute';
export * from './makeTFAttributeKey';
export * from './makeTFArgument';
export * from './makeTFBlockBody';
export * from './makeTFIdentifier';
export * from './makeTFMapExpression';
export * from './makeTFModuleBlock';
export * from './makeTFDynamicBlock';
export * from './makeTFFunctionExpression';
export * from './makeTFResourceBlock';
export * from './makeTFSimpleLiteral';
export * from './makeTFStringArgument';
export * from './makeTFListExpression';
