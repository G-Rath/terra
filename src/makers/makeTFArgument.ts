import { TFArgument, TFLiteralExpression, TFNodeType } from '@src/types';

export const makeTFArgument = <TIdentifier extends string = string>(
  identifier: TIdentifier,
  expression: TFLiteralExpression,
  surroundingText?: Partial<TFArgument['surroundingText']>
): TFArgument<TIdentifier> => ({
  type: TFNodeType.Argument,
  identifier,
  expression,
  surroundingText: {
    leadingInnerText: '',
    trailingInnerText: '',
    ...surroundingText
  }
});
