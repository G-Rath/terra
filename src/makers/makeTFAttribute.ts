import {
  makeTFListExpression,
  makeTFAttributeKey,
  makeTFSimpleLiteral
} from '@src/makers';
import {
  TFIdentifier,
  TFLabel,
  TFLiteralExpression,
  TFNodeType,
  TFAttribute,
  TFAttributeValue
} from '@src/types';

const buildValue = (
  value: TFLiteralExpression | string | Array<TFLiteralExpression | string>
): TFAttributeValue => {
  if (typeof value === 'string') {
    return makeTFSimpleLiteral(value);
  }

  if (Array.isArray(value)) {
    return makeTFListExpression(value);
  }

  return value;
};

export const makeTFAttribute = <TKey extends string = string>(
  key: TKey | TFIdentifier<TKey> | TFLabel<TKey>,
  value: TFAttributeValue | string | Array<TFAttributeValue | string>,
  surroundingText?: Partial<TFAttribute['surroundingText']>
): TFAttribute<TKey> => ({
  type: TFNodeType.Attribute,
  key:
    typeof key === 'string' //
      ? makeTFAttributeKey(key)
      : key,
  value: buildValue(value),
  surroundingText: {
    leadingInnerText: '',
    trailingInnerText: '',
    ...surroundingText
  }
});
