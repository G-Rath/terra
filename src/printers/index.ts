export * from './printBlockBody';
export * from './printBlockLiteral';
export * from './printFileAST';
export * from './printResourceBlock';
