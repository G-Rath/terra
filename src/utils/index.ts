export * from './AwsResourceType';
export * from './asResourceName';
export * from './makeTFArgument';
export * from './makeTFBlockLiteral';
export * from './makeTFDynamicBlock';
export * from './makeTFResourceBlock';
export * from './makeTFStringArgument';
export * from './normaliseRoute53Name';
