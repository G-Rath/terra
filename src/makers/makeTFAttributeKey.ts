import { makeTFIdentifier, makeTFLabel } from '@src/makers';
import { TFAttributeKey } from '@src/types';

export const makeTFAttributeKey = <TKey extends string = string>(
  key: TKey,
  surroundingText?: Partial<TFAttributeKey['surroundingText']>
): TFAttributeKey<TKey> => {
  if (['"', "'"].includes(key[0])) {
    return makeTFLabel(key, surroundingText);
  }

  return makeTFIdentifier(key, surroundingText);
};
