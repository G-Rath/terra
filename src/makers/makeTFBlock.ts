import { makeTFBlockBody } from '@src/makers/makeTFBlockBody';
import { makeTFLabel } from '@src/makers/makeTFLabel';
import {
  TFBlock,
  TFBlockBody,
  TFBlockBodyBody,
  TFLabel,
  TFNodeType,
  TFResourceBlock
} from '@src/types';

export const makeTFBlock = <TIdentifier extends string = string>(
  blockType: string,
  labels: Array<TFLabel | string>,
  body: TFBlockBody<TIdentifier> | TFBlockBodyBody<TIdentifier>,
  surroundingText?: Partial<TFResourceBlock['surroundingText']>
): TFBlock<TIdentifier> => ({
  type: TFNodeType.Block,
  blockType,
  labels: labels.map(label =>
    typeof label === 'string' ? makeTFLabel(label) : label
  ),
  body: Array.isArray(body) ? makeTFBlockBody(body) : body,
  surroundingText: {
    leadingOuterText: '',
    trailingOuterText: '',
    ...surroundingText
  }
});
