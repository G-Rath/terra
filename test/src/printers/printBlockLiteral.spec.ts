import { makeTFArgument } from '@src/makers';
import { printBlockBody, printBlockLiteral } from '@src/printers';
import { TFNodeType } from '@src/types';
import { mocked } from 'ts-jest/utils';

jest.mock('@src/printers/printBlockBody');

describe('printBlockLiteral', () => {
  beforeEach(() =>
    mocked(printBlockBody).mockReturnValue(nameof(printBlockBody))
  );

  it('prints the expression using printBlockBody', () => {
    expect(
      printBlockLiteral({
        type: TFNodeType.Block,
        name: 'ingress',
        body: [makeTFArgument('from_port', 0)]
      })
    ).toMatchSnapshot();
  });

  describe('when printing braces', () => {
    beforeEach(() =>
      mocked(printBlockBody).mockReturnValue(
        ['{', `  ${nameof(printBlockBody)}`, '}'].join('\n')
      )
    );

    it('prints the opening brace on the first line', () => {
      expect(
        printBlockLiteral({
          type: TFNodeType.Block,
          name: 'ingress',
          body: []
        })
      ).toMatchSnapshot();
    });
  });
});
