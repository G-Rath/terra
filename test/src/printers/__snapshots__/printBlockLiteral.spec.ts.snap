// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`printBlockLiteral prints the expression using printBlockBody 1`] = `"ingress printBlockBody"`;

exports[`printBlockLiteral when printing braces prints the opening brace on the first line 1`] = `
"ingress {
  printBlockBody
}"
`;
